Add schemas to support the Plan Builder pages

briri opened the pull request:
## Description

Fixes #119 

- Add `bestPractice` flag to the Section
- Add Project, ProjectContributor, ProjectFunder, ResearchDomain schemas and data-migrations to create tables
- Add Plan, PlanContributor, PlanCollaborator, PlanFeedback, PlanFeedbackComment, PlanAnswer and PlanAnswerComment schemas and data-migrations to create tables

Leaving the work to create the ProjectOutput and PlanOutput schemas/data-migrations as well as the addition of seed data for a separate PR which I will submit tomorrow.

## Type of change
Please delete options that are not relevant

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Include any relevant details for your test configuration.

- Tests pass for the change to the Section. 
- No tests for the schema (`npm run generate` works and I am able to startup the application). 
- I ran the data-migrations in my local and it created the tables and added the bestPractice column to the sections table

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md and added documentation if necessary
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

jupiter007 commented on src/schemas/projectFunder.ts:
```diff
@@ -0,0 +1,92 @@
+    removeProjectFunder(projectFunderId: Int!): Boolean
+
+    "Add a Funder to a Plan"
+    addPlanFunder(planId: Int!, projectFunderId: Int!): PlanFunder
```
L21: I think PlanFunder wasn't defined anywhere

briri replied:
Its just a join table to projectFunders since we don't need need to make any changes, just reference the one(s) we want

briri replied:
maybe makes sense to name the mutation differently? like `selectProjectFunder` or something?

jupiter007 replied:
I'm seeing this error when I start up the app:
```
Attempting to connect to Redis"}
apollo-server   | Error: Unknown type "PlanFunder". Did you mean "Language"?
apollo-server   |     at assertValidSDL (/app/node_modules/graphql/validation/validate.js:135:11)
apollo-server   |     at buildASTSchema (/app/node_modules/graphql/utilities/buildASTSchema.js:44:34)
apollo-server   |     at makeExecutableSchema (/app/node_modules/@apollo/server/node_modules/@graphql-tools/schema/cjs/makeExecutableSchema.js:73:47)
apollo-server   |     at Function.constructSchema (/app/node_modules/@apollo/server/src/ApolloServer.ts:714:32)
apollo-server   |     at new ApolloServer (/app/node_modules/@apollo/server/src/ApolloServer.ts:274:37)
apollo-server   |     at Object.<anonymous> (/app/src/index.ts:28:22)
apollo-server   |     at Module._compile (node:internal/modules/cjs/loader:1480:14)
apollo-server   |     at Module._compile (/app/node_modules/source-map-support/source-map-support.js:521:25)
apollo-server   |     at Module.m._compile (/tmp/ts-node-dev-hook-06277461284618457.js:69:33)
apollo-server   |     at Module._extensions..js (node:internal/modules/cjs/loader:1564:10)
```

briri replied:
Yeah, I just noticed that myself and pushed a fix.

briri replied:
I renamed those mutations to:
```
    "Add a Funder to a Plan"
    selectProjectFunderForPlan(planId: Int!, projectFunderId: Int!): ProjectFunder

    "Remove a PlanFunder from a Plan"
    removeProjectFunderFromPlan(planId: Int!, projectFunderId: Int!): Boolean
```

jupiter007 commented on data-migrations/2024-12-17-1334-create-feedbacks.sql:
```diff
@@ -0,0 +1,32 @@
+# Rounds of administrative feedback for a Data Management Plans (DMPs)
+CREATE TABLE `feedbacks` (
```
Not a big deal, but I think we can just call this table `feedback`, since `feedbacks` is not a word.

briri replied:
Yeah true. haha. I will change it to `feedback`

jupiter007 commented on src/schemas/feedback.ts:
```diff
@@ -0,0 +1,76 @@
+export const typeDefs = gql`
+  extend type Query {
+    "Get all rounds of admin feedback for the plan"
+    planFeedbacks(planId: Int!): [PlanFeedback]
```
L6: Also minor, but we could change `planFeedbacks` to `planFeedback`

jupiter007 commented on src/schemas/researchDomain.ts:
```diff
@@ -0,0 +1,29 @@
+    errors: [String!]
+
+    "The name of the domain"
+    name: String!
```
The ticket has `uri` included as a field, but I don't see it here. Just checking whether it's necessary.

briri replied:
Ah thanks good catch

jupiter007 commented on src/schemas/plan.ts:
```diff
@@ -0,0 +1,97 @@
+    "The plan's answers to the template questions"
+    answers: [Answer!]
+    "Rounds of administrator feedback provided for the Plan"
+    feedbacks: [PlanFeedback!]
```
L95: Could change `feedbacks` here to `feedback` as well.

jupiter007 commented on src/schemas/plan.ts:
```diff
@@ -0,0 +1,97 @@
+    "The last person to have changed any part of the DMP (add collaborators, answer questions, etc.)"
+    lastUpdatedBy: String
+    "The status of the plan"
+    status: PlanStatus
```
Just checking that we don't need the `uploadedNarrativeDoc` here, since it was referenced in the ticket.

I'm assuming `publishedOn` and `publishedBy` and `completedOn` and `completedBy` were not included since we can get that info from `lastUpdatedOn` and `lastUpdatedBy`?

briri replied:
Yeah, those fields will be auto populated based on the current user in `context.token`

jupiter007 commented on src/schemas/project.ts:
```diff
@@ -0,0 +1,36 @@
+import gql from "graphql-tag";
+
```
I didn't find anything on output types. Is that part of another ticket?

briri replied:
going to work on that tomorrow along with the ret of the research outputs

briri commented:
- renamed all the `feedbacks` to `feedback`
- renamed the 2 mutations that let a user add/remove funders from their plan
- Fixed an issue with a bad type reference for `PlanFunder` that was preventing the app from starting up
- added the missing `uri` field to the researchDomain schema

jupiter007 reviewed (approved):
Thanks Brian.

jupiter007 commented:
@briri , actually, I just pulled your branch and saw this error when starting up:
`Error: The type of AddProjectFunderInput.funder must be Input Type but got: Affiliation!.`

briri commented:
sry. I wish `npm run generate` would catch stuff like that. It should startup now

jupiter007 commented:
> sry. I wish `npm run generate` would catch stuff like that. It should startup now

Your changes look good to me. Thanks @briri
